File: arrays/segmented-sieve.py
import sys; sys.stdin = open(sys.path[0] + '\\input.txt', 'r'); sys.stdout = open(sys.path[0] + '\\output.txt', 'w')


def simple_sieve():
    mark = [False] * (n + 1)

    for i in range(3, n + 1, 2):
        if not mark[i]:
            primes.append(i)
            j = i * i
            while j <= n:
                mark[i] = True
                j += (2 * i)


def segmented_sieve(a, b):
    mark = [False] * n
    nsqrt = int(b ** 0.5)
    if a == 1:
        a += 1
    
    for p in primes:
        if p > nsqrt:
            break
        j = p * p  
        if j < a:
            j = ((a + p - 1) // p) * p
        
        while j <= b:
            mark[j - a] = True
            j += p

    for i in range(a, b + 1):
        if not mark[i - a]:
            print(i)


n = 10 ** 6 + 1
primes = [2]
simple_sieve()

for tc in range(int(input())):
    a, b = list(map(int, input().split()))
    segmented_sieve(a, b)

File: arrays/segmented-sieve.cpp
#include <iostream>
#include <algorithm>
#include <string>
#include <vector>
#include <cmath>

#define ll long long
#define watch(x) cout << (#x) << " is " << (x) << endl

using namespace std;


const int s = (int) 1e6 + 1;
bool mark[s];
vector<ll> primes;


void simple_sieve() {
    fill(mark, mark + s, false);
    primes.push_back(2);

    for (ll i = 3; i <= s; i += 2) {
        if (!mark[i]) {
            primes.push_back(i);
            for (ll j = i * i; j <= s; j += (2 * i)) {
                mark[j] = true;
            }
        }
    }
}


void segmented_sieve(ll a, ll b) {
    fill(mark, mark + s, false);

    ll nsqrt = sqrt(b);
    if (a == 1) a++;

    for (ll i = 0; i < primes.size() && primes[i] <= nsqrt; i++) {
        ll p = primes[i];
        ll j = p * p;
        if (j < a) j = ((a + p - 1) / p) * p;

        for (; j <= b; j += p) {
            mark[j - a] = true;
        }
    }

    for (ll i = a; i <= b; i++) {
        if (!mark[i - a]) {
            cout << i << '\n';
        }
    }
}

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);

    #ifndef ONLINE_JUDGE
    freopen("input.txt", "r", stdin);
    freopen("output.txt", "w", stdout);
    #endif

    ll tc, a, b;
    cin >> tc;

    simple_sieve();

    while (tc--) {
        cin >> a >> b;
        segmented_sieve(a, b);
        cout << '\n';
    }

    return 0;
}